model/project_ref.go

```diff
@@ -2434,17 +2434,17 @@ func GetMessageForPatch(patchID string) (string, error) {
 	return project.CommitQueue.Message, nil
 }
 
-func ValidateContainers(project string, containers []Container) error {
-	pRef, err := FindMergedProjectRef(project, "", true)
-	if err != nil {
-		return errors.Errorf("Unable to find merged project ref for project %s", project)
-	}
+// ValidateContainers inspects the list of containers defined in the project YAML and checks that each
+// are properly configured, and that their definitions can coexist with what is defined for container sizes
+// on the project admin page.
+func ValidateContainers(pRef *ProjectRef, containers []Container) error {
 	catcher := grip.NewSimpleCatcher()
 	for _, container := range containers {
-		catcher.Add(ValidateContainerSystem(container.System))
-		catcher.Add(ValidateContainerSize(container.Resources))
-		catcher.Add(ValidateContainerSize(pRef.ContainerSizes[container.Size]))
-		catcher.NewWhen(container.Size != "" && pRef.ContainerSizes[container.Size] == nil, fmt.Sprintf("size '%s' is not defined anywhere", container.Size))
+		catcher.Add(container.System.Validate())
+		catcher.Add(container.Resources.Validate())
+		size, ok := pRef.ContainerSizes[container.Size]
+		catcher.AddWhen(ok, size.Validate())
```

Comment:
I would check `ok` separately because when written like this, Go will execute `size.Validate()` regardless of whether `ok` is true or false:
```go
if ok {
    catcher.Add(size.Validate())
}
```

Small example might explain why better than my words:
```go
import (
	"fmt"

	"github.com/mongodb/grip"
)

func foo() string {
	fmt.Println("hi, don't print me")
	return "hi"
}

func main() {
	m := map[string]struct{}{}
	m["apple"] = struct{}{}
	catcher := grip.NewBasicCatcher()
	_, ok := m["banana"]
	catcher.NewWhen(ok, foo()) // foo still executes, so it prints "hi, don't print me"
	fmt.Println("should not have added any errors:", catcher.Resolve())
}
````